EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+---
+eip: 9999
+title: Block-level warming
+description: Warm addresses and storage keys over the duration of a block
+author: Toni Wahrstätter (@nerolation), Alex Stokes (@ralexstokes), Ansgar Dietrichs (@adietrichs)
+discussions-to: ""
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-01-15
+---
+
+## Abstract
+
+This proposal introduces block-level address and storage key warming, allowing accessed addresses and storage keys to maintain their "warm" status throughout an entire block's execution. Accessed slots can be effectively cached at the block level, allowing for this opimization.
+
+## Motivation
+
+Currently, the EVM's storage slot warming mechanism operates at the transaction level, requiring each transaction to "warm up" slots independently, even when accessing the same storage locations within the same block. This design does not take advantage of the fact that modern node implementations can effectively cache storage access patterns at the block level. By extending the slot warming duration to the block level, we can:
+
+1. Reduce redundant warming costs for frequently accessed slots
+2. Better align gas costs with actual computational overhead
+3. Improve overall network throughput without compromising security.
+
+As of Jan. 2025, the empirically messured efficiency gains are around 5%.
+
+
+## Specification
+
+### Mechanics
+
+When a storage slot is accessed within a block:
+1. The first access to a slot in a block incurs the cold access cost as of [EIP-2929](./eip-2929).
```

Comment:
```suggestion
1. The first access to a slot in a block incurs the cold access cost as of [EIP-2929](./eip-2929.md).
```

Reply:
Except if it has been added to the access list. This raises the question also what to do with the access lists: if a tx has an address/storage slot in the access list, do we put it at the warm access of the block level before executing that transaction? (So, even if the transaction does not access the slot/address at all, is this still added to the block warm addresses/slots?)

Reply:
Never mind, from the code below it is clear these should be added.

Reply:
yeah, this can be further clarified in the EIP. Access lists would then warm for all consecutive txs too.